typescript/infra/src/tx/transaction-reader.ts

```diff
@@ -0,0 +1,454 @@
+import { Result } from '@ethersproject/abi';
+import { decodeMultiSendData } from '@safe-global/protocol-kit/dist/src/utils/index.js';
+import {
+  MetaTransactionData,
+  OperationType,
+} from '@safe-global/safe-core-sdk-types';
+import { BigNumber, ethers } from 'ethers';
+
+import {
+  AnnotatedEV5Transaction,
+  ChainMap,
+  ChainName,
+  CoreConfig,
+  EvmIsmReader,
+  HyperlaneReader,
+  InterchainAccount,
+  MultiProvider,
+  coreFactories,
+  interchainAccountFactories,
+  normalizeConfig,
+} from '@hyperlane-xyz/sdk';
+import {
+  addressToBytes32,
+  bytes32ToAddress,
+  deepEquals,
+  eqAddress,
+  retryAsync,
+} from '@hyperlane-xyz/utils';
+
+import {
+  icaOwnerChain,
+  icas,
+  safes,
+} from '../../config/environments/mainnet3/owners.js';
+import { DeployEnvironment } from '../config/environment.js';
+import { getSafeAndService } from '../utils/safe.js';
+
+export class TransactionReader extends HyperlaneReader {
+  errors: any[] = [];
+
+  constructor(
+    readonly environment: DeployEnvironment,
+    readonly multiProvider: MultiProvider,
+    readonly chain: ChainName,
+    readonly chainAddresses: ChainMap<Record<string, string>>,
+    readonly coreConfig: ChainMap<CoreConfig>,
+  ) {
+    super(multiProvider, chain);
+  }
+
+  async read(chain: ChainName, tx: AnnotatedEV5Transaction): Promise<any> {
+    try {
+      return await this.doRead(chain, tx);
+    } catch (e) {
+      console.error('Error reading transaction', e, chain, tx);
+      throw e;
+    }
+  }
+
+  async doRead(chain: ChainName, tx: AnnotatedEV5Transaction): Promise<any> {
+    // If it's to an ICA
+    if (this.isIcaTransaction(chain, tx)) {
+      return this.readIcaTransaction(chain, tx);
+    }
+
+    // If it's to a Mailbox
+    if (this.isMailboxTransaction(chain, tx)) {
+      return this.readMailboxTransaction(chain, tx);
+    }
+
+    if (await this.isMultisendTransaction(chain, tx)) {
+      return this.readMultisendTransaction(chain, tx);
+    }
+
+    return {};
+  }
+
+  private async readIcaTransaction(
+    chain: ChainName,
+    tx: AnnotatedEV5Transaction,
+  ): Promise<any> {
+    if (!tx.data) {
+      console.log('No data in ICA transaction');
+      return undefined;
+    }
+    const { symbol } = await this.multiProvider.getNativeToken(chain);
+    const decoded =
+      interchainAccountFactories.interchainAccountRouter.interface.parseTransaction(
+        {
+          data: tx.data,
+          value: tx.value,
+        },
+      );
+
+    const args = formatFunctionFragmentArgs(
+      decoded.args,
+      decoded.functionFragment,
+    );
+    let prettyArgs = args;
+    if (decoded.functionFragment.name === 'enrollRemoteRouters') {
```

Comment:
nit: can we import something that would avoid comparing to a literal?

Reply:
yeah just moved to something type safe on the interface